### PR TITLE
instant server; wait_user_init; protocol_attach; switch_to_generated_user; bugfix

### PR DESCRIPTION
**Pull request by stneng**

### CODE REVIEW

File: examples/protocol_greetings_with_test.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            .await?;
+        Ok(())
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    let is0 = InstantServer::new();
+    let is1 = InstantServer::new();
+    let cl0 = is0.to_colink().clone_and_switch_to_generated_user().await?;
```
Comment (camelop) [inline examples/protocol_greetings_with_test.rs:38]: get_colink

File: examples/protocol_greetings_with_test.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    let cl1 = is0.to_colink().clone_and_switch_to_generated_user().await?;
+    cl0.wait_user_init().await?;
```
Comment (camelop) [inline examples/protocol_greetings_with_test.rs:40]: can we merge this into `clone_and_switch_to_generated_user`

File: src/extensions/instant_server.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                .unwrap();
+        Self {
+            id: instant_server_id,
+            port,
+            host_token,
+            process: child,
+        }
+    }
+
+    pub fn to_colink(&self) -> CoLink {
```
Comment (camelop) [inline src/extensions/instant_server.rs:102]: func name

File: src/extensions/instant_server.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            .stderr(Stdio::null())
+            .spawn()
+            .unwrap();
+        loop {
+            if std::fs::metadata(working_dir.join("host_token.txt")).is_ok()
+                && std::net::TcpStream::connect(&format!("127.0.0.1:{}", port)).is_ok()
+            {
+                break;
+            }
+            std::thread::sleep(core::time::Duration::from_millis(100));
```
Comment (camelop) [inline src/extensions/instant_server.rs:88]: can you measure this?

File: src/extensions/instant_server.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        let host_token: String =
+            String::from_utf8_lossy(&std::fs::read(working_dir.join("host_token.txt")).unwrap())
+                .parse()
+                .unwrap();
+        Self {
```
Comment (camelop) [inline src/extensions/instant_server.rs:94]: can we make sure the server is up before returning? (probably need to make the function async)

**Review by camelop — CHANGES_REQUESTED** [review]
LG! Plz see the detailed comments

File: src/extensions/instant_server.rs (hunk 4 above)
Comment (stneng) [reply src/extensions/instant_server.rs:88]: result

Comment (camelop) [reply src/extensions/instant_server.rs:88]: Let's use 10ms then. Later we can probably automate this as a test case.

File: examples/protocol_greetings_with_test.rs (hunk 2 above)
Comment (camelop) [reply examples/protocol_greetings_with_test.rs]: or maybe we can just keep `switch_to_generated_user`

**Review by camelop — APPROVED** [review]
Looks great!

**Review by camelop — APPROVED** [review]

**Pull request merged**